compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -957,34 +957,45 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
     ) -> SubstsRef<'tcx> {
         let generics = self.tcx.generics_of(def_id);
         let mut num_supplied_defaults = 0;
-        let mut type_params = generics
-            .params
-            .iter()
-            .rev()
-            .filter_map(|param| match param.kind {
-                ty::GenericParamDefKind::Lifetime => None,
-                ty::GenericParamDefKind::Type { has_default, .. } => {
-                    Some((param.def_id, has_default))
-                }
-                // FIXME(const_generics:defaults)
-                ty::GenericParamDefKind::Const { has_default: _has_default } => None,
-            })
-            .peekable();
-        let has_default = {
-            let has_default = type_params.peek().map(|(_, has_default)| has_default);
-            *has_default.unwrap_or(&false)
-        };
-        if has_default {
-            let types = substs.types().rev();
-            for ((def_id, has_default), actual) in type_params.zip(types) {
-                if !has_default {
-                    break;
+
+        #[derive(PartialEq, Eq, Copy, Clone)]
+        enum Kind {
+            Const,
+            Type,
+        }
+        let default_params = generics.params.iter().rev().filter_map(|param| match param.kind {
+            ty::GenericParamDefKind::Type { has_default: true, .. } => {
+                Some((param.def_id, Kind::Type))
+            }
+            ty::GenericParamDefKind::Const { has_default: true } => {
+                Some((param.def_id, Kind::Const))
+            }
+            _ => None,
+        });
+        let mut types = substs.types().rev();
```

Comment:
The defaults should be provided in the correct order, which means we should be able to just use `substs.rev()` here, and match on the kind of the `Subst`, right?